File: 09-backtracking/00078-Subsets/00078-subsets.cpp
// 78: Subsets
// https://leetcode.com/problems/subsets/


#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    void backtrack(vector<int>& nums, int start, vector<int>& current, vector<vector<int>>& result) {
        result.push_back(current);

        for (int i=start; i<nums.size(); i++) {
            current.push_back(nums[i]);
            backtrack(nums, i + 1, current, result);
            current.pop_back();
        }
    }

public:
    // SOLUTOIN
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> current;
        vector<vector<int>> result;
        backtrack(nums, 0, current, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {1,2,3};

    // OUTPUT
    auto result = o.subsets(nums);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}

File: 09-backtracking/00078-Subsets/00078-subsets.py
# 78: Subsets
# https://leetcode.com/problems/subsets/

from typing import List

class Solution:
    # SOLUTION
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        current = []

        def backtrack(start: int) -> None:
            result.append(current.copy())

            for i in range(start , len(nums)):
                current.append(nums[i])
                backtrack(i+1)
                current.pop()

        backtrack(0)
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums = [1,2,3]

    # OUTPUT
    result = o.subsets(nums)

    print(result)
